rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/internal_Node.hpp

```diff
@@ -66,21 +69,44 @@
 namespace rmf_traffic_ros2 {
 namespace schedule {
 
+using namespace std::chrono_literals;
+
 //==============================================================================
 class ScheduleNode : public rclcpp::Node
 {
 public:
+  using QueryMap =
+    std::unordered_map<uint64_t, rmf_traffic::schedule::Query>;
+  using QuerySubscriberCountMap =
+    std::unordered_map<uint64_t, uint64_t>;
+
+  ScheduleNode(
+    std::shared_ptr<rmf_traffic::schedule::Database> database_,
+    QueryMap registered_queries_,
+    const rclcpp::NodeOptions& options = rclcpp::NodeOptions());
 
   ScheduleNode(const rclcpp::NodeOptions& options);
 
   ~ScheduleNode();
 
+  virtual void setup(
+    QuerySubscriberCountMap registered_query_subscriber_counts);
+
+  std::chrono::milliseconds heartbeat_period = 1s;
+  rclcpp::QoS heartbeat_qos_profile;
+  using Heartbeat = rmf_traffic_msgs::msg::Heartbeat;
+  using HeartbeatPub = rclcpp::Publisher<Heartbeat>;
+  HeartbeatPub::SharedPtr heartbeat_pub;
+
+  virtual void setup_redundancy();
+  virtual void start_heartbeat();
+
   using request_id_ptr = std::shared_ptr<rmw_request_id_t>;
 
   using RegisterQuery = rmf_traffic_msgs::srv::RegisterQuery;
   using RegisterQueryService = rclcpp::Service<RegisterQuery>;
 
-  void register_query(
+  virtual void register_query(
```

Comment:
I understand the motive for using virtual functions here (makes testing much easier), but it violates the principle of [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance).

I think it's fine to let it pass in this PR, but let's leave a quality improvement issue ticket to do a refactor here when time permits.

That refactor could be achieved by making separate classes/structs for setting up and managing these different groups of services, and then sharing those structs between this official node and the other test nodes.

Reply:
That sounds good to me. Most of the `rmf_ros2` stuff needs refactoring for easier testing, so I'll create a general ticket.

Reply:
[Issue created](https://github.com/open-rmf/rmf_ros2/issues/73).